Kernel/VirtIO/VirtIOConsolePort.cpp

```diff
@@ -0,0 +1,166 @@
+        auto remaining_space = m_receive_buffer->bytes_till_end();
+
+        // Our algorithm always has only one buffer in the queue.
+        VERIFY(popped_chain.length() == 1);
+        VERIFY(!queue.new_data_available());
+        popped_chain.release_buffer_slots_to_queue();
+
+        VirtIOQueueChain new_chain(queue);
+        if (remaining_space != 0) {
+            new_chain.add_buffer_to_chain(used_space.offset(used), remaining_space, BufferType::DeviceWritable);
+            m_console.supply_chain_and_notify(m_receive_queue, new_chain);
+        } else {
+            m_receive_buffer_exhausted = true;
+        }
+
+        evaluate_block_conditions();
+    } else {
+        ScopedSpinLock ringbuffer_lock(m_transmit_buffer->lock());
+        auto& queue = m_console.get_queue(m_transmit_queue);
+        ScopedSpinLock queue_lock(queue.lock());
+        size_t used;
+        VirtIOQueueChain popped_chain = queue.pop_used_buffer_chain(used);
+        do {
+            popped_chain.for_each([this](PhysicalAddress address, size_t length) {
+                m_transmit_buffer->reclaim_space(address, length);
+            });
+            popped_chain.release_buffer_slots_to_queue();
+            popped_chain = queue.pop_used_buffer_chain(used);
+        } while (!popped_chain.is_empty());
+        // Unblock any IO tasks that were blocked because can_write() returned false
+        evaluate_block_conditions();
+    }
+}
+
+bool VirtIOConsolePort::can_read(const FileDescription&, size_t) const
+{
+    return m_receive_buffer->used_bytes() > 0;
+}
+
+KResultOr<size_t> VirtIOConsolePort::read(FileDescription& desc, u64, UserOrKernelBuffer& buffer, size_t size)
+{
+    if (!size)
+        return 0;
+
+    if (!can_read(desc, size))
+        return EAGAIN;
+
+    ScopedSpinLock ringbuffer_lock(m_receive_buffer->lock());
+
+    auto bytes_copied = m_receive_buffer->copy_data_out(size, buffer).value();
+    m_receive_buffer->reclaim_space(m_receive_buffer->start_of_used(), bytes_copied);
+
+    if (m_receive_buffer_exhausted && m_receive_buffer->used_bytes() == 0) {
+        auto& queue = m_console.get_queue(m_receive_queue);
+        ScopedSpinLock queue_lock(queue.lock());
+        VirtIOQueueChain new_chain(queue);
+        new_chain.add_buffer_to_chain(m_receive_buffer->start_of_region(), RINGBUFFER_SIZE, BufferType::DeviceWritable);
+        m_console.supply_chain_and_notify(m_receive_queue, new_chain);
+        m_receive_buffer_exhausted = false;
+    }
+
+    return bytes_copied;
+}
+
+bool VirtIOConsolePort::can_write(const FileDescription&, size_t) const
+{
+    return m_console.get_queue(m_transmit_queue).has_free_slots() && m_transmit_buffer->has_space();
+}
+
+KResultOr<size_t> VirtIOConsolePort::write(FileDescription& desc, u64, const UserOrKernelBuffer& data, size_t size)
+{
+    if (!size)
+        return 0;
+
+    if (!can_write(desc, size))
+        return EAGAIN;
+
+    ScopedSpinLock ringbuffer_lock(m_transmit_buffer->lock());
```

Comment:
This lock acquire might need to be acquired before checking can_write(), otherwise we have a TOCTTOU (where we get past the can_write, then switch to another thread before grabbing the lock).